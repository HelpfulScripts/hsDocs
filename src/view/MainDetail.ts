import { m, Vnode}              from 'hslayout';
import { Layout }               from 'hslayout';
import { DocSets }              from '../DocSets'; 
import { comment, commentLong } from './MainComment';
import { flags, sourceLink, 
         signature, type, 
         extensionOf, inheritedFrom,
         kindString, itemName, 
         makeID, 
         implementationOf,
         extendedBy,
         implementedBy}      from './Parts'; 


/**
 * Creates Documentation on the main panel 
 */
export class MainDetail extends Layout { 
    getComponents(node:Vnode): Vnode {
        const lib =  m.route.param('lib');
        const field = m.route.param('field');
        node.attrs.route = undefined;

        let result = getOverview(lib, field) || itemDoc(DocSets.get(lib, field) || ''); 
        return m('.hs-main-detail', [result]); 
    }
}

/**
 * Checks if the project overview is being requested and returns the overview, 
 * or `undefined` if not available
 * @param mdl the module name to check
 * @return Vnode containing the overview file, or `undefined`
 */
function getOverview(lib:string, mdl:string):Vnode {
    if (mdl === '0' || mdl === '') {  //show module overview
        mdl = DocSets.get(lib, `${lib}.overview`);
        if (mdl) { // if project has an overview:
            return overviewDoc(mdl); 
        }
    }
    return undefined;
}

/**
 * Creates documentation for standard items in the main panel
 * @param mdl the module to document on the main panel
 */
function itemDoc(mdl:any) {
    const sig = mdl.signatures? mdl.signatures[0] : mdl;
    return m('.hs-item-doc', [
        title(mdl, sig),
        commentLong(sig),
        members(mdl, sig)
    ]);
}

/**
 * Creates documentation for the project overview in the main panel
 * @param mdl the module to document on the main panel
 */
function overviewDoc(mdl:any) {
    const sig = mdl.signatures? mdl.signatures[0] : mdl;
    return m('.hs-item-doc', [
        commentLong(sig),
    ]);
}

/**
 * renders the title of the main panel.
 * ![title image](data/title.png)
 * @param mdl the module to document 
 * @param sig a signature of the module, or the the module itself
 */
function title(mdl:any, sig:any): Vnode { 
    return m('.hs-item-title', {id: makeID('title', mdl)}, itemDescriptor(mdl, sig)); 
}

function members(mdl:any, sig:any): Vnode {
    if (mdl.groups) {
        return m('.hs-item-members', [
            ...mdl.groups.map((g:any) => member(g, mdl.lib, true, true)),   // static public
            ...mdl.groups.map((g:any) => member(g, mdl.lib, true, false)),  // static protected and private
            ...mdl.groups.map((g:any) => member(g, mdl.lib, false, true)),  // instance public
            ...mdl.groups.map((g:any) => member(g, mdl.lib, false, false))  // instance protected and private
        ]);
    } else if (mdl.parameters) {
        return m('.hs-item-members', parameter(mdl.parameters, mdl.lib));
    } else if (mdl.type && mdl.type.declaration) {
        mdl.type.declaration.lib = mdl.lib;
        return members(mdl.type.declaration, sig);
    } else {
        return m('.hs-item-members');
    }
}

function parameter(g:any[], lib:string): Vnode {
    let content = g.map((c:any) => m('.hs-item-parameter', {id:makeID('parameter', c)}, itemChild(c)));
    content.unshift(m('.hs-item-member-title', {id:'parameters'}, m('span', 'Parameters')));
    return m('.hs-item-member', content);
}

function member(group:any, lib:string, statc:boolean, publc: boolean): Vnode {
    const resolve           = ((c:number) => DocSets.get(lib, c));
    const directChildren    = ((mdl:any) => !mdl['inheritedFrom']);
    const inheritedChildren = ((mdl:any) =>  mdl['inheritedFrom']);
    const groupMap = {
        'External modules': '.hs-item-external-module',
        'Constructors':     '.hs-item-constructor',
        'Classes':          '.hs-item-class',          
        'Interfaces':       '.hs-item-interface',          
        'Functions':        '.hs-item-function',          
        'Methods':          '.hs-item-method',          
        'Variables':        '.hs-item-variable',
        'Object literals':  '.hs-item-object-literal',
        'Properties':       '.hs-item-property',
        'Enumerations':     '.hs-item-enumration',
        'Type aliases':     '.hs-item-alias',          
        'Accessors':        '.hs-item-accessors'        
    };
    const fn = groupMap[group.title] || '.hs-item-unknown-member';
    const isPublic = (flags:any) => flags.isPublic || (flags.isExported && !flags.isPrivate && !flags.isProtected);

    const content = group.children
        .map(resolve)
        .filter(directChildren)
        .filter((mdl:any) => statc? mdl.flags.isStatic : !mdl.flags.isStatic)
        .filter((mdl:any) => publc? isPublic(mdl.flags) : !isPublic(mdl.flags))
        .map((mdl:any) => m(fn, {id:makeID(group.title, mdl)}, itemChild(mdl)));
    const inherited = group.children
        .map(resolve)
        .filter(inheritedChildren)
        .filter((mdl:any) => statc? mdl.flags.isStatic : !mdl.flags.isStatic)
        .filter((mdl:any) => publc? mdl.flags.isPublic : !mdl.flags.isPublic)
        .map((mdl:any) => m(`.hs-item-inherited ${fn}`, {id:makeID(group.title, mdl)}, itemChild(mdl)));

    const publStr = publc?'Public':'Protected or Private';
    const statStr = statc?'Static':''; 
    if (inherited.length>0) {
        inherited.unshift(m('.hs-item-inherited .hs-item-member-title', m('span', `${publStr} ${statStr} Inherited ${group.title}`)));
    }
    if (content.length>0) {
        content.unshift(m('.hs-item-member-title', {id:group.title.toLowerCase()}, m('span', `${publStr} ${statStr} ${group.title}`)));
    }
    return m(`.hs-item-member ${statc?'.hs-item-static':''} ${publc?'.hs-item-public':''}`, content.concat(inherited));
}

/**
 * prints the descriptor line for a specified signature.
 * @param mdl 
 * @param sig 
 */
function itemDescriptor(mdl:any, sig:any):Vnode {
    try { return m('.hs-item-desc', [ 
            flags(mdl),
            kindString(mdl),
            itemName(mdl, mdl),
            signature(sig, mdl),
            type(sig,  mdl.lib),
            extensionOf(mdl),
            implementationOf(mdl),
            inheritedFrom(mdl),
            sourceLink(mdl),
            extendedBy(mdl),
            implementedBy(mdl)
        ]);
    }
    catch(e) { console.log(e); console.log(mdl); }
}

function itemChild(mdl:any, sig=mdl): Vnode[] {
    return mdl.signatures? 
        mdl.signatures.map((s:any) => m('.hs-item-child-signature',[itemDescriptor(mdl, s), comment(s, true)])) : 
        [itemDescriptor(mdl, sig), comment(sig,true)];
}


