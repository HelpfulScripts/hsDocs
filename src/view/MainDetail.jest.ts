import { MainDetail }   from './MainDetail';
import m                from "mithril";
import { Log }          from 'hsnode'; const log = new Log('MainDetail.jest');
import { DocSets }      from '../DocSets';
import { libLinkByPath } from '../NodesDisplay';
const fs = require('fs');


const set = {
    "docs": [
        "hsDocs.json",
        "hsGraphD3.json",
        "hsWidget.json",
        "hsNode.json",
        "hsUtil.json"
    ],
    "title": "HS Libraries"
};

m.request = (req: any) => {
    if (req.url === 'mylist.json') {
        return Promise.resolve(set);
    } else if(req.url.endsWith('.json')) {
        return new Promise((resolve:(data:any)=>void, reject:(err:any)=>void) => {
            fs.readFile(`${__dirname}/../test/${req.url}`, 'utf8', (err:any, data:any) => {
                if (err) { throw err; }
                else { resolve(data); }
            });
        })
        .then((data:any) => (typeof data === 'string')? JSON.parse(data) : data)
        .catch(console.log)
        ;
    } else {
        return Promise.resolve('dir');
    }
};

const route = {
    field: '0',
    lib: 'hsWidget'
};

m.route.param = (name?:string):string => {
    if (name==='field') { return route.field; }
    if (name==='lib')   { return route.lib; }
    return '';
};


describe('MainDetail', () => {
    describe('overview', () => {
        const root:any = window.document.createElement("body");
        beforeAll(async (done) => { 
            m.route.get = () => '/api/hsDocs/0';
            await DocSets.loadList('mylist.json')
            route.field = '0';
            // route.field = 'hsGraphD3.plots.Pie.Pie';
            m.mount(root, { view: () => m(MainDetail, {lib:route.lib, field:route.field}) }); 
            done();
        });
    
        it('exists', (done)=>{
            expect(MainDetail).toBeDefined();
            done();
        });
    
        it('matches snapshot', (done)=>{
            expect(root).toMatchSnapshot();
            done();
        });
    });

    describe('hsDocs.DocSets', () => {
        const root:any = window.document.createElement("body");
        beforeAll( async (done) => { 
            m.route.get = () => '/api/hsDocs/hsDocs.DocSets';
            await DocSets.loadList('mylist.json')
            route.field = 'hsDocs.DocSets';
            m.mount(root, { view: () => m(MainDetail, {lib:route.lib, field:route.field}) }); 
            done();
        });
    
        it('exists', (done)=>{
            expect(MainDetail).toBeDefined();
            done();
        });
    
        it('matches snapshot', (done)=>{
            expect(root).toMatchSnapshot();
            done();
        });
    });

    describe('hsDocs.DocSets.DocSets.add', () => {
        const root:any = window.document.createElement("body");
        beforeAll(async (done) => { 
            m.route.get = () => '/api/hsDocs/hsDocs.DocSets.DocSets.add';
            await DocSets.loadList('mylist.json')
            route.field = 'hsDocs.DocSets.DocSets.add';
            m.mount(root, { view: () => { return m(MainDetail, {lib:route.lib, field:route.field});} }); 
            done();
        });
    
        it('exists', (done)=>{
            expect(MainDetail).toBeDefined();
            done();
        });
    
        it('matches snapshot', (done)=>{
            expect(root).toMatchSnapshot();
            done();
        });
    });

    describe('hsGraphD3.plots.Pie.Pie', () => {
        const root:any = window.document.createElement("body");
        beforeAll(async (done) => { 
            m.route.get = () => '/api/hsGraphD3/hsGraphD3.plots.Pie.Pie';
            await DocSets.loadList('mylist.json')
            route.lib = 'hsGraphD3';
            route.field = 'hsGraphD3.plots.Pie.Pie';
            m.mount(root, { view: () => { return m(MainDetail, {lib:route.lib, field:route.field});} }); 
            done();
        });
    
        it('exists', (done)=>{
            expect(MainDetail).toBeDefined();
            done();
        });
    
        it('matches snapshot', (done)=>{
            expect(root).toMatchSnapshot();
            done();
        });
    });
});
