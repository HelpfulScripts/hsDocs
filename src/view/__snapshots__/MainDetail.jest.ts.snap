// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainDetail hsDocs.DocSets matches snapshot 1`] = `
<body>
  <div
    class="hs-layout  "
  >
    <div
      class="hsdocs"
    >
      <div
        class="hsdocs_title"
        id="title_docsets"
      >
        <span
          class="hsdocs_flags"
        >
          <span
            class="hsdocs_flag_export"
          >
            export
          </span>
        </span>
        <span
          class="hsdocs_kind"
        >
          module
        </span>
        <span
          class="hsdocs_itemname"
        >
          <a
            href="#!/api/hsDocs/hsDocs.DocSets"
          >
            DocSets
          </a>
        </span>
        <span
          class="hsdocs_signatures"
        />
        <span
          class="hsdocs_title_colon"
        >
          : 
        </span>
        <span>
          
        </span>
        <span
          class="hsdocs_extensionOf"
        />
        <span
          class="hsdocs_inheritedFrom"
        />
        <span
          class="hsdocs_source"
        >
          <a
            href="data/src/hsDocs/DocSets.html#0"
            target="_blank"
          >
            [source]
          </a>
        </span>
      </div>
      <div
        class="hsdocs_head_comment"
      >
        <span
          class="hsdocs_comment_desc"
        >
          <p>
            DocSets.ts. Loads the doc.json files to process and display as documentation.
Processing occurs in these steps:
          </p>
          

          <ol>
            

            <li>
              Load the index.json file that contains a list of doc.json files to load, one for each library to show
            </li>
            

            <li>
              Load each doc.json file, which describes a library
            </li>
            

            <li>
              Call DocSets.add to add the library name to the registry and create an index of entries for the library
            </li>
            

          </ol>
        </span>
        <div />
      </div>
      <div
        class="hsdocs_members"
      >
        <div
          class="hsdocs_member  hsdocs_flag_public"
        >
          <div
            class="hsdocs_member_title"
          >
            Exported Classes
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.DocSets"
          >
            <div
              class="hsdocs_child"
            >
              <span>
                <span
                  class="hsdocs_flags"
                >
                  <span
                    class="hsdocs_flag_export"
                  >
                    export
                  </span>
                </span>
                <span
                  class="hsdocs_kind"
                >
                  class
                </span>
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.DocSets"
                  >
                    DocSets
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                />
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  <a
                    href="data/src/hsDocs/DocSets.html#33"
                    target="_blank"
                  >
                    [source]
                  </a>
                </span>
              </span>
              <div
                class="hsdocs_sig_comment"
              >
                <span
                  class="hsdocs_comment_desc"
                >
                  <p>
                    DocSets object. Keeps a list of registered docsets and
provides access to elements of each docset.
                  </p>
                </span>
                <div />
              </div>
            </div>
          </div>
        </div>
        <div
          class="hsdocs_member  hsdocs_flag_public"
        >
          <div
            class="hsdocs_member_title"
          >
            Exported Type aliases
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.json"
          >
            <div
              class="hsdocs_child"
            >
              <span>
                <span
                  class="hsdocs_flags"
                >
                  <span
                    class="hsdocs_flag_export"
                  >
                    export
                  </span>
                </span>
                <span
                  class="hsdocs_kind"
                >
                  type
                </span>
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.json"
                  >
                    json
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                />
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_intrinsic"
                    >
                      any
                    </span>
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  <a
                    href="data/src/hsDocs/DocSets.html#17"
                    target="_blank"
                  >
                    [source]
                  </a>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="hsdocs_member "
        >
          <div
            class="hsdocs_member_title"
          >
            Internal Variables
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.DOCDIR"
          >
            <div
              class="hsdocs_child"
            >
              <span>
                <span
                  class="hsdocs_flags"
                >
                  <span
                    class="hsdocs_flag_const"
                  >
                    const
                  </span>
                </span>
                
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.DOCDIR"
                  >
                    DOCDIR
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                />
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_intrinsic"
                    >
                      string
                    </span>
                  </span>
                  <span
                    class="hsdocs_type_default"
                  >
                     = "./data/"
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  <a
                    href="data/src/hsDocs/DocSets.html#9"
                    target="_blank"
                  >
                    [source]
                  </a>
                </span>
              </span>
            </div>
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.FILE"
          >
            <div
              class="hsdocs_child"
            >
              <span>
                <span
                  class="hsdocs_flags"
                >
                  <span
                    class="hsdocs_flag_const"
                  >
                    const
                  </span>
                </span>
                
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.FILE"
                  >
                    FILE
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                />
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_intrinsic"
                    >
                      string
                    </span>
                  </span>
                  <span
                    class="hsdocs_type_default"
                  >
                     = "./data/index.json"
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  <a
                    href="data/src/hsDocs/DocSets.html#15"
                    target="_blank"
                  >
                    [source]
                  </a>
                </span>
              </span>
              <div
                class="hsdocs_sig_comment"
              >
                <span
                  class="hsdocs_comment_desc"
                >
                  <p>
                    the default location for the index .json files, relative to the web page:

                    <code>
                      './data/index.json'
                    </code>
                  </p>
                </span>
                <div />
              </div>
            </div>
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.log"
          >
            <div
              class="hsdocs_child"
            >
              <span>
                <span
                  class="hsdocs_flags"
                >
                  <span
                    class="hsdocs_flag_const"
                  >
                    const
                  </span>
                </span>
                
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.log"
                  >
                    log
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                />
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_reference"
                    >
                      Log
                    </span>
                  </span>
                  <span
                    class="hsdocs_type_default"
                  >
                     =  new Log('DocsSets')
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  <a
                    href="data/src/hsDocs/DocSets.html#6"
                    target="_blank"
                  >
                    [source]
                  </a>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="hsdocs_member "
        >
          <div
            class="hsdocs_member_title"
          >
            Internal Functions
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.loadDocSet"
          >
            <div
              class="hsdocs_child hsdocs_signature"
            >
              <span>
                <span
                  class="hsdocs_flags"
                />
                <span
                  class="hsdocs_kind"
                >
                  function
                </span>
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.loadDocSet"
                  >
                    loadDocSet
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                >
                  <span
                    class="hsdocs_itemname"
                  >
                     (
                  </span>
                  <span>
                    
                    <span
                      class="hsdocs_signature_param"
                    >
                      <span
                        class="hsdocs_itemname"
                      >
                        dir
                      </span>
                      <span
                        class="hsdocs_colon"
                      >
                        :
                      </span>
                      <span>
                        <span
                          class="hsdocs_type"
                        >
                          <span
                            class="hsdocs_type_intrinsic"
                          >
                            string
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                  <span>
                    , 
                    <span
                      class="hsdocs_signature_param"
                    >
                      <span
                        class="hsdocs_itemname"
                      >
                        file
                      </span>
                      <span
                        class="hsdocs_colon"
                      >
                        :
                      </span>
                      <span>
                        <span
                          class="hsdocs_type"
                        >
                          <span
                            class="hsdocs_type_intrinsic"
                          >
                            string
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                  <span
                    class="hsdocs_itemname"
                  >
                    )
                  </span>
                </span>
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_reference"
                    >
                      Promise
                      &lt;
                      <span
                        class="hsdocs_type_intrinsic"
                      >
                        void
                      </span>
                      &gt;
                    </span>
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  
                </span>
              </span>
              <div
                class="hsdocs_childsig_comment"
              >
                <span
                  class="hsdocs_comment_desc"
                >
                  <p>
                    Loads a docset specified by file from the directory 
                    <code>
                      dir
                    </code>
                    .
Once received, the docset is registered in 
                    <code>
                      modules
                    </code>
                     via the 
                    <code>
                      add
                    </code>
                     method.
                  </p>
                </span>
                <div
                  class="hsdocs_comment_params"
                >
                  <div
                    class="hsdocs_params_title"
                  >
                    parameters:
                  </div>
                  <div
                    class="hsdocs_comment_param"
                  >
                    <span
                      class="hsdocs_param_name"
                    >
                      dir
                      : 
                    </span>
                    <span
                      class="hsdocs_param_comment"
                    >
                      <span
                        class="hsdocs_comment_desc"
                      >
                        <p>
                          the directory to read from
                        </p>
                      </span>
                      <div />
                    </span>
                  </div>
                  <div
                    class="hsdocs_comment_param"
                  >
                    <span
                      class="hsdocs_param_name"
                    >
                      file
                      : 
                    </span>
                    <span
                      class="hsdocs_param_comment"
                    >
                      <span
                        class="hsdocs_comment_desc"
                      >
                        <p>
                          the 
                          <code>
                            json
                          </code>
                           file to load as docset
                        </p>
                      </span>
                      <div />
                    </span>
                  </div>
                </div>
                <div />
              </div>
            </div>
          </div>
          <div
            class="hsdocs_node"
            id="hsDocs.DocSets.matchAll"
          >
            <div
              class="hsdocs_child hsdocs_signature"
            >
              <span>
                <span
                  class="hsdocs_flags"
                />
                <span
                  class="hsdocs_kind"
                >
                  function
                </span>
                <span
                  class="hsdocs_itemname"
                >
                  <a
                    href="#!/api/hsDocs/hsDocs.DocSets.matchAll"
                  >
                    matchAll
                  </a>
                </span>
                <span
                  class="hsdocs_signatures"
                >
                  <span
                    class="hsdocs_itemname"
                  >
                     (
                  </span>
                  <span>
                    
                    <span
                      class="hsdocs_signature_param"
                    >
                      <span
                        class="hsdocs_itemname"
                      >
                        str
                      </span>
                      <span
                        class="hsdocs_colon"
                      >
                        :
                      </span>
                      <span>
                        <span
                          class="hsdocs_type"
                        >
                          <span
                            class="hsdocs_type_intrinsic"
                          >
                            string
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                  <span>
                    , 
                    <span
                      class="hsdocs_signature_param"
                    >
                      <span
                        class="hsdocs_itemname"
                      >
                        re
                      </span>
                      <span
                        class="hsdocs_colon"
                      >
                        :
                      </span>
                      <span>
                        <span
                          class="hsdocs_type"
                        >
                          <span
                            class="hsdocs_type_reference"
                          >
                            RegExp
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                  <span
                    class="hsdocs_itemname"
                  >
                    )
                  </span>
                </span>
                <span
                  class="hsdocs_title_colon"
                >
                  : 
                </span>
                <span>
                  <span
                    class="hsdocs_type"
                  >
                    <span
                      class="hsdocs_type_array"
                    >
                      Array&lt;
                      <span
                        class="hsdocs_type_intrinsic"
                      >
                        string
                      </span>
                      &gt;
                    </span>
                  </span>
                </span>
                <span
                  class="hsdocs_extensionOf"
                />
                <span
                  class="hsdocs_inheritedFrom"
                />
                <span
                  class="hsdocs_source"
                >
                  
                </span>
              </span>
              <div
                class="hsdocs_childsig_comment"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MainDetail hsDocs.DocSets.DocSets.add matches snapshot 1`] = `
<body>
  <div
    class="hs-layout  "
  >
    <div
      class="hsdocs"
    >
      no item
    </div>
  </div>
</body>
`;

exports[`MainDetail hsGraphD3.plots.Pie.Pie matches snapshot 1`] = `
<body>
  <div
    class="hs-layout  "
  >
    <div
      class="hsdocs"
    >
      no item
    </div>
  </div>
</body>
`;

exports[`MainDetail overview matches snapshot 1`] = `
<body>
  <div
    class="hs-layout  "
  >
    <div
      class="hsdocs"
    >
      <div
        class="hsdocs_head_comment"
      >
        <span
          class="hsdocs_comment_desc"
        >
          <h1
            id="hsgraphd3"
          >
            hsGraphD3
          </h1>
          

          <p>
            Helpful Scripts Graph convenience wrapper for D3js.

            <a
              href="./data/src/hsGraphD3/coverage/"
              target="_blank"
            >
              [Coverage Info]
            </a>
          </p>
          

          <p>
            <a
              href="https://badge.fury.io/js/hsgraphd3"
            >
              <img
                alt="npm version"
                src="https://badge.fury.io/js/hsgraphd3.svg"
              />
            </a>
            

            <a
              href="https://github.com/helpfulscripts/hsgraphd3"
            >
              <img
                alt="GitHub"
                src="https://img.shields.io/badge/GitHub-hsGraphD3-blue.svg"
              />
            </a>
            

            <a
              href="https://helpfulscripts.github.io/hsGraphD3/#!/api/hsGraphD3/0"
            >
              <img
                alt="docs"
                src="https://img.shields.io/badge/hsDocs-hsGraphD3-blue.svg"
              />
            </a>
            

            <a
              href="https://travis-ci.org/HelpfulScripts/hsGraphD3"
            >
              <img
                alt="Build Status"
                src="https://travis-ci.org/HelpfulScripts/hsGraphD3.svg?branch=master"
              />
            </a>
            

            <a
              href="https://david-dm.org/helpfulscripts/hsgraphd3"
            >
              <img
                alt="Dependencies Status"
                src="https://david-dm.org/helpfulscripts/hsgraphd3.svg"
              />
            </a>
            

            <a
              href="https://coveralls.io/github/HelpfulScripts/hsGraphD3?branch=master"
            >
              <img
                alt="Coverage Status"
                src="https://coveralls.io/repos/github/HelpfulScripts/hsGraphD3/badge.svg?branch=master"
              />
            </a>
            

            <a
              href="https://snyk.io/test/github/HelpfulScripts/hsGraphD3?targetFile=package.json"
            >
              <img
                alt="Known Vulnerabilities"
                src="https://snyk.io/test/github/HelpfulScripts/hsGraphD3/badge.svg?targetFile=package.json"
              />
            </a>
            

            <a
              href="https://www.npmjs.com/package/hsgraphd3"
            >
              <img
                alt="NPM License"
                src="https://img.shields.io/badge/license-MIT-brightgreen.svg"
              />
            </a>
          </p>
          

          <hr />
          

          <p>
            <code>
              hsGraphD3
            </code>
             provides convienent programmatic shortcuts to plotting data with 
            <a
              href="https://d3js.org/"
            >
              <code>
                D3js
              </code>
            </a>
            .
          </p>
          

          <p>
            See 
            <a
              href="#!/api/hsDocs/hsDocs.API"
            >
              API
            </a>
            .
          </p>
          

          <h2
            id="usage"
          >
            Usage
          </h2>
          

          <ol>
            

            <li>
              create a 
              <a
                href="#!/api/hsDocs/hsDocs.Graph"
              >
                <code>
                  Graph
                </code>
              </a>
               object with a root DOM element to attach to:

              <code>
                graph = new hsGraphD3.GraphCartesian(root);
              </code>
            </li>
            

            <li>
              add desired series configurations, specifying the type of series, as well as the data to use for the x- and y-axes,
and other variables, depending on the series type:

              <code>
                graph.series.add('area', {x:'time', y:'volume', r:()=&gt;5});
              </code>
            </li>
            

            <li>
              apply any desired formatting changes to the default:

              <code>
                graph.axes.defaults.color = '#88f';
              </code>
            </li>
            

            <li>
              render the graph for a given data set:

              <code>
                graph.render(data);
              </code>
            </li>
            

          </ol>
          

          <p>
            To dynamically update the graph to new data, call the 
            <a
              href="#!/api/hsDocs/hsDocs.Graph.RenderChain"
            >
              <code>
                update
              </code>
            </a>
             method.
See 
            <a
              href="#!/api/hsDocs/hsDocs.examples"
            >
              examples
            </a>
            .
          </p>
          

          <h2
            id="data"
          >
            Data
          </h2>
          

          <p>
            Data is provided either in table form
          </p>
          

          <pre>
            <code>
              [
   colNames[],
   &lt;number|string|Date&gt;[],
   ...
]

            </code>
          </pre>
          

          <p>
            or in object-literal form
          </p>
          

          <pre>
            <code>
              {
   name?:  string;
   colNames:  string[];
   rows:   &lt;number|string|Date&gt;[][];
}

            </code>
          </pre>
          

          <h2
            id="example-1-link-plotsbubble-bubble-chart"
          >
            Example 1: 
            <a
              href="#!/api/hsDocs/hsDocs.plots.Bubble"
            >
              <code>
                bubble
              </code>
            </a>
             chart
          </h2>
          

          <style />
          

          <p>
            <example
              class="hs-layout-frame"
              id="hs100000"
              style="height:200px"
            />
          </p>
          

          <h2
            id="example-2-link-plotsline-line-vs-link-plotstimeseries-timeseries"
          >
            Example 2: 
            <a
              href="#!/api/hsDocs/hsDocs.plots.Line"
            >
              <code>
                line
              </code>
            </a>
             vs. 
            <a
              href="#!/api/hsDocs/hsDocs.plots.TimeSeries"
            >
              <code>
                timeseries
              </code>
            </a>
          </h2>
          

          <style />
          

          <p>
            <example
              class="hs-layout-frame"
              id="hs100002"
              style="height:200px"
            />
          </p>
        </span>
        <div />
      </div>
    </div>
  </div>
</body>
`;
